File: internal_testsuites/golang/testsuite/stream_logs_test/stream_logs_test.go
package stream_logs_test

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/testsuite/consts"
	"github.com/kurtosis-tech/kurtosis/api/golang/core/lib/services"
	"github.com/kurtosis-tech/kurtosis/api/golang/engine/lib/kurtosis_context"
	"github.com/stretchr/testify/require"
	"testing"
	"time"
)

const (
	testName = "stream-logs"

	exampleServiceName services.ServiceName = "stream-logs"

	testTimeOut = 180 * time.Second

	shouldFollowLogs    = true
	shouldNotFollowLogs = false

	nonExistentServiceUuid = "stream-logs-1667939326-non-existent"

	firstLogLine  = "kurtosis"
	secondLogLine = "test"
	thirdLogLine  = "running"
	lastLogLine   = "successfully"

	// Retry configuration for log retrieval
	// Logs may not be immediately available due to Fluentbit/Vector flush timing;
	// in CI the flush can take well over 30 seconds so we allow up to ~100 s of retries.
	maxLogRetrievalRetries    = 10
	logRetrievalRetryInterval = 10 * time.Second
)

var (
	doNotFilterLogLines   *kurtosis_context.LogLineFilter = nil
	doesContainTextFilter                                 = kurtosis_context.NewDoesContainTextLogLineFilter(lastLogLine)

	exampleServiceLogLines = []string{
		firstLogLine,
		secondLogLine,
		thirdLogLine,
		lastLogLine,
	}

	logLinesByService = map[services.ServiceName][]string{
		exampleServiceName: exampleServiceLogLines,
	}
)

type serviceLogsRequestInfoAndExpectedResults struct {
	requestedEnclaveIdentifier   string
	requestedServiceUuids        map[services.ServiceUUID]bool
	requestedFollowLogs          bool
	expectedLogLines             []string
	expectedNotFoundServiceUuids map[services.ServiceUUID]bool
	logLineFilter                *kurtosis_context.LogLineFilter
}

func TestStreamLogs(t *testing.T) {

	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, _, destroyEnclaveFunc, err := test_helpers.CreateEnclave(t, ctx, testName)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer func() {
		err = destroyEnclaveFunc()
		require.NoError(t, err, "An error occurred destroying the enclave after the test finished")
	}()

	// ------------------------------------- TEST SETUP ----------------------------------------------
	kurtosisCtx, err := kurtosis_context.NewKurtosisContextFromLocalEngine()
	require.NoError(t, err)

	serviceList, err := test_helpers.AddServicesWithLogLines(ctx, enclaveCtx, logLinesByService)
	require.NoError(t, err, "An error occurred adding the datastore service")

	// It takes some time for logs to persist so we sleep to ensure logs have persisted
	// Otherwise the test is flaky
	time.Sleep(consts.FluentbitRefreshInterval)
	// ------------------------------------- TEST RUN ----------------------------------------------

	enclaveUuid := enclaveCtx.GetEnclaveUuid()

	serviceUuids := map[services.ServiceUUID]bool{}
	for _, serviceCtx := range serviceList {
		serviceUuid := serviceCtx.GetServiceUUID()
		serviceUuids[serviceUuid] = true
	}

	serviceLogsRequestInfoAndExpectedResultsList := getServiceLogsRequestInfoAndExpectedResultsList(
		string(enclaveUuid),
		serviceUuids,
	)

	for _, serviceLogsRequestInfoAndExpectedResultsObj := range serviceLogsRequestInfoAndExpectedResultsList {

		requestedEnclaveIdentifier := serviceLogsRequestInfoAndExpectedResultsObj.requestedEnclaveIdentifier
		requestedServiceUuids := serviceLogsRequestInfoAndExpectedResultsObj.requestedServiceUuids
		requestedShouldFollowLogs := serviceLogsRequestInfoAndExpectedResultsObj.requestedFollowLogs
		expectedLogLines := serviceLogsRequestInfoAndExpectedResultsObj.expectedLogLines
		expectedNonExistenceServiceUuids := serviceLogsRequestInfoAndExpectedResultsObj.expectedNotFoundServiceUuids
		filter := serviceLogsRequestInfoAndExpectedResultsObj.logLineFilter

		expectedLogLinesByService := map[services.ServiceUUID][]string{}
		for userServiceUuid := range requestedServiceUuids {
			expectedLogLinesByService[userServiceUuid] = expectedLogLines
		}

		// Retry log retrieval to handle Fluentbit/Vector flush timing
		var receivedLogLinesByService map[services.ServiceUUID][]string
		var receivedNotFoundServiceUuids map[services.ServiceUUID]bool
		var testEvaluationErr error
		var logsRetrieved bool

		for attempt := 0; attempt < maxLogRetrievalRetries; attempt++ {
			receivedLogLinesByService, receivedNotFoundServiceUuids, testEvaluationErr = test_helpers.GetLogsResponse(
				t,
				ctx,
				testTimeOut,
				kurtosisCtx,
				requestedEnclaveIdentifier,
				requestedServiceUuids,
				expectedLogLinesByService,
				requestedShouldFollowLogs,
				filter,
			)

			if testEvaluationErr != nil {
				t.Logf("Attempt %d: error retrieving logs: %v", attempt+1, testEvaluationErr)
				time.Sleep(logRetrievalRetryInterval)
				continue
			}

			// Check if we received enough logs for all services
			logsRetrieved = true
			for userServiceUuid := range requestedServiceUuids {
				expectedLines := expectedLogLinesByService[userServiceUuid]
				receivedLines := receivedLogLinesByService[userServiceUuid]
				if len(receivedLines) < len(expectedLines) {
					logsRetrieved = false
					t.Logf("Attempt %d: expected %d log lines for service %s, got %d. Retrying...",
						attempt+1, len(expectedLines), userServiceUuid, len(receivedLines))
					break
				}
			}

			if logsRetrieved {
				break
			}
			time.Sleep(logRetrievalRetryInterval)
		}

		require.True(t, logsRetrieved || len(expectedLogLines) == 0,
			"Failed to retrieve expected logs after %d attempts", maxLogRetrievalRetries)
		require.NoError(t, testEvaluationErr)

		for userServiceUuid := range requestedServiceUuids {
			expectedLines := expectedLogLinesByService[userServiceUuid]
			receivedLines := receivedLogLinesByService[userServiceUuid]
			require.GreaterOrEqual(t, len(receivedLines), len(expectedLines),
				"Expected at least %d log lines for service %s, but got %d",
				len(expectedLines), userServiceUuid, len(receivedLines))
			for logNum, expectedLogLine := range expectedLines {
				require.Contains(t, receivedLines[logNum], expectedLogLine)
			}
		}
		require.Equal(t, expectedNonExistenceServiceUuids, receivedNotFoundServiceUuids)
	}
}

// ====================================================================================================
//
//	Private helper functions
//
// ====================================================================================================
func getServiceLogsRequestInfoAndExpectedResultsList(
	enclaveIdentifier string,
	serviceUuids map[services.ServiceUUID]bool,
) []*serviceLogsRequestInfoAndExpectedResults {

	emptyServiceUuids := map[services.ServiceUUID]bool{}

	nonExistentServiceUuids := map[services.ServiceUUID]bool{
		nonExistentServiceUuid: true,
	}

	firstCallRequestInfoAndExpectedResults := &serviceLogsRequestInfoAndExpectedResults{
		requestedEnclaveIdentifier:   enclaveIdentifier,
		requestedServiceUuids:        serviceUuids,
		requestedFollowLogs:          shouldFollowLogs,
		expectedLogLines:             []string{lastLogLine},
		expectedNotFoundServiceUuids: emptyServiceUuids,
		logLineFilter:                doesContainTextFilter,
	}

	secondCallRequestInfoAndExpectedResults := &serviceLogsRequestInfoAndExpectedResults{
		requestedEnclaveIdentifier:   enclaveIdentifier,
		requestedServiceUuids:        serviceUuids,
		requestedFollowLogs:          shouldFollowLogs,
		expectedLogLines:             []string{firstLogLine, secondLogLine, thirdLogLine, lastLogLine},
		expectedNotFoundServiceUuids: emptyServiceUuids,
		logLineFilter:                doNotFilterLogLines,
	}

	thirdCallRequestInfoAndExpectedResults := &serviceLogsRequestInfoAndExpectedResults{
		requestedEnclaveIdentifier:   enclaveIdentifier,
		requestedServiceUuids:        serviceUuids,
		requestedFollowLogs:          shouldNotFollowLogs,
		expectedLogLines:             []string{firstLogLine, secondLogLine, thirdLogLine, lastLogLine},
		expectedNotFoundServiceUuids: emptyServiceUuids,
		logLineFilter:                doNotFilterLogLines,
	}

	fourthCallRequestInfoAndExpectedResults := &serviceLogsRequestInfoAndExpectedResults{
		requestedEnclaveIdentifier:   enclaveIdentifier,
		requestedServiceUuids:        nonExistentServiceUuids,
		requestedFollowLogs:          shouldFollowLogs,
		expectedLogLines:             []string{},
		expectedNotFoundServiceUuids: nonExistentServiceUuids,
		logLineFilter:                doNotFilterLogLines,
	}

	serviceLogsRequestInfoAndExpectedResultsList := []*serviceLogsRequestInfoAndExpectedResults{
		firstCallRequestInfoAndExpectedResults,
		secondCallRequestInfoAndExpectedResults,
		thirdCallRequestInfoAndExpectedResults,
		fourthCallRequestInfoAndExpectedResults,
	}

	return serviceLogsRequestInfoAndExpectedResultsList
}

File: internal_testsuites/typescript/src/testsuite/stream_log_test/stream_logs.test.ts
import {
    EnclaveUUID,
    KurtosisContext,
    LogLineFilter,
    ServiceContext,
    ServiceUUID,
    ServiceName,
    ServiceLog,
} from "kurtosis-sdk";
import log from "loglevel";
import {err, Result} from "neverthrow";
import {createEnclave} from "../../test_helpers/enclave_setup";
import {addServicesWithLogLines, getLogsResponseAndEvaluateResponse} from "../../test_helpers/test_helpers";

const TEST_NAME = "stream-logs";
const IS_PARTITIONING_ENABLED = false;

const EXAMPLE_SERVICE_NAME: ServiceName = "stream-logs";

const SHOULD_FOLLOW_LOGS = true;
const SHOULD_NOT_FOLLOW_LOGS = false;

const NON_EXISTENT_SERVICE_UUID = "stream-logs-1667939326-non-existent";

const FIRST_LOG_LINE_STR = "kurtosis"
const SECOND_LOG_LINE_STR = "test"
const THIRD_LOG_LINE_STR = "running"
const LAST_LOG_LINE_STR = "successfully"

const FIRST_LOG_LINE = new ServiceLog(FIRST_LOG_LINE_STR)
const SECOND_LOG_LINE = new ServiceLog(SECOND_LOG_LINE_STR)
const THIRD_LOG_LINE = new ServiceLog(THIRD_LOG_LINE_STR)
const LAST_LOG_LINE = new ServiceLog(LAST_LOG_LINE_STR)

const DO_NOT_FILTER_LOG_LINES = undefined;
const DOES_CONTAIN_TEXT_FILTER = LogLineFilter.NewDoesContainTextLogLineFilter(LAST_LOG_LINE_STR);

const EXAMPLE_SERVICE_LOG_LINES = [FIRST_LOG_LINE, SECOND_LOG_LINE, THIRD_LOG_LINE, LAST_LOG_LINE];

const LOG_LINES_BY_SERVICE = new Map<ServiceName, ServiceLog[]>([
    [EXAMPLE_SERVICE_NAME, EXAMPLE_SERVICE_LOG_LINES],
])

class ServiceLogsRequestInfoAndExpectedResults {
    readonly requestedEnclaveUUID: EnclaveUUID;
    readonly requestedServiceUuids: Set<ServiceUUID>;
    readonly requestedFollowLogs: boolean;
    readonly expectedLogLines: ServiceLog[];
    readonly expectedNotFoundServiceUuids: Set<ServiceUUID>;
    readonly logLineFilter: LogLineFilter | undefined;

    constructor(
        requestedEnclaveUUID: EnclaveUUID,
        requestedServiceUuids: Set<ServiceUUID>,
        requestedFollowLogs: boolean,
        expectedLogLines: ServiceLog[],
        expectedNotFoundServiceUuids: Set<ServiceUUID>,
        logLineFilter: LogLineFilter | undefined
    ) {
        this.requestedEnclaveUUID = requestedEnclaveUUID;
        this.requestedServiceUuids = requestedServiceUuids;
        this.requestedFollowLogs = requestedFollowLogs;
        this.expectedLogLines = expectedLogLines;
        this.expectedNotFoundServiceUuids = expectedNotFoundServiceUuids;
        this.logLineFilter = logLineFilter;
    }
}

jest.setTimeout(180000);

test("Test Stream Logs", TestStreamLogs);

async function TestStreamLogs() {

    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED);

    if (createEnclaveResult.isErr()) {
        throw createEnclaveResult.error;
    }

    const {enclaveContext, stopEnclaveFunction} = createEnclaveResult.value;

    try {
        // ------------------------------------- TEST SETUP ----------------------------------------------
        const newKurtosisContextResult = await KurtosisContext.newKurtosisContextFromLocalEngine();
        if (newKurtosisContextResult.isErr()) {
            log.error(`An error occurred connecting to the Kurtosis engine for running test stream logs`);
            return err(newKurtosisContextResult.error);
        }
        const kurtosisCtx = newKurtosisContextResult.value;

        const serviceListResult: Result<Map<ServiceName, ServiceContext>, Error> = await addServicesWithLogLines(enclaveContext, LOG_LINES_BY_SERVICE);

        if (serviceListResult.isErr()) {
            throw new Error(`An error occurred adding the services for the test. Error:\n${serviceListResult.error}`);
        }

        const serviceList: Map<ServiceName, ServiceContext> = serviceListResult.value;

        if (LOG_LINES_BY_SERVICE.size != serviceList.size) {
            throw new Error(`Expected number of added services '${LOG_LINES_BY_SERVICE.size}', but the actual number of added services is '${serviceList.size}'`);
        }

        // ------------------------------------- TEST RUN ----------------------------------------------

        const enclaveID: EnclaveUUID = enclaveContext.getEnclaveUuid();

        const serviceUuids: Set<ServiceUUID> = new Set<ServiceUUID>();

        for (let [, serviceCtx] of serviceList) {
            const serviceUuid = serviceCtx.getServiceUUID();
            serviceUuids.add(serviceUuid);
        }

        const serviceLogsRequestInfoAndExpectedResultsList = getServiceLogsRequestInfoAndExpectedResultsList(
            enclaveID,
            serviceUuids,
        )

        for (let serviceLogsRequestInfoAndExpectedResults of serviceLogsRequestInfoAndExpectedResultsList) {

            const requestedEnclaveUUID = serviceLogsRequestInfoAndExpectedResults.requestedEnclaveUUID;
            const requestedServiceUuids = serviceLogsRequestInfoAndExpectedResults.requestedServiceUuids;
            const requestedShouldFollowLogs = serviceLogsRequestInfoAndExpectedResults.requestedFollowLogs;
            const expectedLogLines = serviceLogsRequestInfoAndExpectedResults.expectedLogLines;
            const expectedNonExistenceServiceUuids = serviceLogsRequestInfoAndExpectedResults.expectedNotFoundServiceUuids;
            const filter = serviceLogsRequestInfoAndExpectedResults.logLineFilter;

            let expectedLogLinesByService: Map<ServiceUUID, ServiceLog[]> = new Map<ServiceUUID, ServiceLog[]>;
            for (const userServiceUuid of requestedServiceUuids) {
                expectedLogLinesByService.set(userServiceUuid, expectedLogLines);
            }

            const getLogsResponseResult = await getLogsResponseAndEvaluateResponse(
                kurtosisCtx,
                requestedEnclaveUUID,
                requestedServiceUuids,
                expectedLogLinesByService,
                expectedNonExistenceServiceUuids,
                requestedShouldFollowLogs,
                filter,
            )

            if (getLogsResponseResult.isErr()) {
                throw getLogsResponseResult.error
            }
        }
    } finally {
        stopEnclaveFunction();
    }

    jest.clearAllTimers();
}

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================
function getServiceLogsRequestInfoAndExpectedResultsList(
    enclaveID: EnclaveUUID,
    serviceUuids: Set<ServiceUUID>,
): Array<ServiceLogsRequestInfoAndExpectedResults> {

    const emptyServiceUuids: Set<ServiceUUID> = new Set<ServiceUUID>();
    const nonExistentServiceUuids: Set<ServiceUUID> = new Set<ServiceUUID>();
    nonExistentServiceUuids.add(NON_EXISTENT_SERVICE_UUID);

    const firstCallRequestInfoAndExpectedResults: ServiceLogsRequestInfoAndExpectedResults = new ServiceLogsRequestInfoAndExpectedResults(
        enclaveID,
        serviceUuids,
        SHOULD_FOLLOW_LOGS,
        [LAST_LOG_LINE],
        emptyServiceUuids,
        DOES_CONTAIN_TEXT_FILTER,
    )

    const secondCallRequestInfoAndExpectedResults: ServiceLogsRequestInfoAndExpectedResults = new ServiceLogsRequestInfoAndExpectedResults(
        enclaveID,
        serviceUuids,
        SHOULD_FOLLOW_LOGS,
        [FIRST_LOG_LINE, SECOND_LOG_LINE, THIRD_LOG_LINE, LAST_LOG_LINE],
        emptyServiceUuids,
        DO_NOT_FILTER_LOG_LINES,
    )

    const thirdCallRequestInfoAndExpectedResults: ServiceLogsRequestInfoAndExpectedResults = new ServiceLogsRequestInfoAndExpectedResults(
        enclaveID,
        serviceUuids,
        SHOULD_NOT_FOLLOW_LOGS,
        [FIRST_LOG_LINE, SECOND_LOG_LINE, THIRD_LOG_LINE, LAST_LOG_LINE],
        emptyServiceUuids,
        DO_NOT_FILTER_LOG_LINES,
    )

    const fourthCallRequestInfoAndExpectedResults: ServiceLogsRequestInfoAndExpectedResults = new ServiceLogsRequestInfoAndExpectedResults(
        enclaveID,
        nonExistentServiceUuids,
        SHOULD_FOLLOW_LOGS,
        [],
        nonExistentServiceUuids,
        DO_NOT_FILTER_LOG_LINES,
    )

    const serviceLogsRequestInfoAndExpectedResultsList: Array<ServiceLogsRequestInfoAndExpectedResults> = new Array<ServiceLogsRequestInfoAndExpectedResults>();
    serviceLogsRequestInfoAndExpectedResultsList.push(firstCallRequestInfoAndExpectedResults)
    serviceLogsRequestInfoAndExpectedResultsList.push(secondCallRequestInfoAndExpectedResults)
    serviceLogsRequestInfoAndExpectedResultsList.push(thirdCallRequestInfoAndExpectedResults)
    serviceLogsRequestInfoAndExpectedResultsList.push(fourthCallRequestInfoAndExpectedResults)

    return serviceLogsRequestInfoAndExpectedResultsList
}
